chia/wallet/puzzles/aggregator.clvm

```diff
@@ -0,0 +1,133 @@
+            (merge-lists (r morphed_conditions) conditions)
+        )
+    )
+
+    (defun aggregate-solutions (INNER_AGG_PUZZLE_HASH AMOUNT solutions create_coin_condition conditions)
+        (if solutions
+            (if (= (sha256tree1 (f (f solutions))) INNER_AGG_PUZZLE_HASH)
+                (morph-and-aggregate
+                    (morph-conditions (a (f (f solutions)) (f (r (f solutions))))
+                            AMOUNT
+                            create_coin_condition
+                            conditions)
+                    AMOUNT
+                    solutions
+                    conditions
+                )
+                ;(list (morph-conditions (a (f (f solutions)) (f (r (f solutions)))) AMOUNT create_coin_condition conditions)); solutions conditions)
+                (x "bad inner puzzle")
+            )
+            ; return back create coin that will be used to curry new puzzle hash for aggregator
+            (c create_coin_condition conditions)
+        )
+    )
+
+    (defun morph-conditions (conditions singleton_amount odd_create_coin morphed_conditions)
+        (if conditions
+            (if (all (= (f (f conditions)) CREATE_COIN)); (= (f (r (r (f conditions)))) singleton_amount) (not odd_create_coin))
+                ; it's a create odd coin, add it to main params
+                (morph-conditions
+                    (r conditions)
+                    singleton_amount
+                    (f conditions)
+                    morphed_conditions
+                )
+;                (if (all (= (f conditions) CREATE_PUZZLE_ANNOUNCEMENT) (= (substr (f (r (conditions))) 0 3) POSITION_ANN_PREFIX))
+;                    ; ignore position puzzle announcements
+;                    (morph-conditions (r conditions) singleton_amount odd_create_coin morphed_conditions)
+                    (morph-conditions (r conditions) singleton_amount odd_create_coin (c (f conditions) morphed_conditions))
+;                )
+            )
+            (if odd_create_coin
+                (c odd_create_coin morphed_conditions)
+                (x "wrong conditions")
+            )
+        )
+    )
+
+    (defun recreate (aggregated_conditions MOD_HASH MOD_LOCK_HASH)
+        (c
+            (c CREATE_COIN
+              (c
+                    (get-new-puzzle-hash
+                        MOD_HASH
+                        MOD_LOCK_HASH
+                        (f (r (f aggregated_conditions))) ; new agg puzzle hash
+                    )
+                    (r (r (f aggregated_conditions)))
+              )
+            )
+            ; conditions
+           (r aggregated_conditions)
+        )
+    )
+    (merge-lists
+        (recreate
+            (aggregate-solutions
+                INNER_AGG_PUZZLE_HASH amount solutions () ())
+            MOD_HASH
+            MOD_LOCK_HASH
+        )
+        (ensure-order MOD_HASH MOD_LOCK_HASH solutions 1)
```

Comment:
I'm not sure I follow exactly what `ensure-order` does.  If I understand correctly, the idea here is to do multiple spends of the same singleton all together in one spend.  There's no intermediary steps on chain before all of the states get processed so why are we creating all of these coins and consuming them?  The solution value `solutions` is malleable by the mempool and that solution value is the determination of the order so the order can be manipulated no matter what, right?

Reply:
It's using replace bundle feature in mempool that requires at least all spends to be in the new bundle as old one. 
So aggregator creates lock coins for each solution with solution index in the list. This coin needs to be spend when solution is added as aggregator asserts announcement from it and lock coin asserts announcement from aggregator. 
The part I haven't added yet is that to replace the bundle it also needs to have higher fee per cost than previous one. This requires that each new solution spends a coin to pay for transaction where it could also assert that lock coin has been spent. This makes it impossible for farmer to mess with the bundle and solutions. 
It's possible to also add agg sig me in the lock coin, maybe that's better.